urld0f681c558_farom57_AstroidDriver_tgz-pJ8-laazotea_indi_server_examples_MinimumINDIServerJ8/src/laazotea/indi/driver/examples/INDIDriverExample.java

```diff
@@ -0,0 +1,156 @@
+import laazotea.indi.INDIException;
+import laazotea.indi.driver.*;
+
+/**
+ * An example class representing a very basic INDI Driver.
+ *
+ * @author S. Alonso (Zerjillo) [zerjioi at ugr.es]
+ * @version 1.3, April 5, 2012
+ */
+public class INDIDriverExample extends INDIDriver implements INDIConnectionHandler {
+
+  // The Properties and Elements of this Driver
+  private INDIBLOBProperty imageP;
+  private INDIBLOBElement imageE;
+  private INDISwitchProperty sendP;
+  private INDISwitchElement sendE;
+
+  public INDIDriverExample(InputStream inputStream, OutputStream outputStream) {
+    super(inputStream, outputStream);
+
+    // Define the BLOB Property with this Driver as its owner, name "image", label "Image", group "Image Properties", initial state IDLE and Read Only.
+    imageP = new INDIBLOBProperty(this, "image", "Image", "Image Properties", PropertyStates.IDLE, PropertyPermissions.RO);
+    // Define the BLOB Element with name "image" and label "Image". Its initial value is empty.
+    imageE = new INDIBLOBElement(imageP, "image", "Image");
+
+
+    // Define the Switch Property with this driver as its owner, name "sendImage", label "Send Image", group "Image Properties", initial state IDLE, Read/Write permission and AtMostOne rule for the switch.
+    sendP = new INDISwitchProperty(this, "sendImage", "Send Image", "Image Properties", PropertyStates.IDLE, PropertyPermissions.RW, SwitchRules.AT_MOST_ONE);
+    // Define the Switch Element with name "sendImage", label "Send Image" and initial status OFF
+    sendE = new INDISwitchElement(sendP, "sendImage", "Send Image", SwitchStatus.OFF);
+  }
+
+  // Returns the name of the Driver
+  @Override
+  public String getName() {
+    return "INDI Driver Example";
+  }
+
+  // No Number Properties can be set by clients. Empty method  
+  @Override
+  public void processNewNumberValue(INDINumberProperty property, Date timestamp, INDINumberElementAndValue[] elementsAndValues) {
+  }
+
+  // No BLOB Properties can be set by clients. Empty method  
+  @Override
+  public void processNewBLOBValue(INDIBLOBProperty property, Date timestamp, INDIBLOBElementAndValue[] elementsAndValues) {
+  }
+
+  // No Text Properties can be set by clients. Empty method
+  @Override
+  public void processNewTextValue(INDITextProperty property, Date timestamp, INDITextElementAndValue[] elementsAndValues) {
+  }
+
+  // Processes the changes sent by the client to the Switch Properties. If the Switch property is the CONNECTION one it adds or removes the image and send Properties. If the Property is the "sendImage", it loads an image from disk, sets it to the image property and sends it back to the client.
+  @Override
+  public void processNewSwitchValue(INDISwitchProperty property, Date timestamp, INDISwitchElementAndValue[] elementsAndValues) {
+
+    if (property == sendP) {  // If the property is the sendImage one
+      if (elementsAndValues.length > 0) {  // If any element has been updated
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;elementsAndValues&#39; (index: 2 | type: INDISwitchElementAndValue[). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;property&#39;. The same parameter type is validated here for example: urld0f681c558_farom57_AstroidDriver_tgz-pJ8-laazotea_indi_server_examples_MinimumINDIServerJ8/src/farom/astroiddriver/INDIAstroidDriver.java:816. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).